src/proxy_lib/proxy_lib.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <umf/memory_pool.h>
+#include <umf/memory_provider.h>
+#include <umf/providers/provider_os_memory.h>
+
+#include "../utils/utils_common.h"
+#include "proxy_lib.h"
+#include "umf/pools/pool_jemalloc.h"
+
+static umf_memory_provider_handle_t OS_memory_provider = NULL;
+static umf_memory_pool_handle_t Proxy_pool = NULL;
+
+// it protects us from recursion in umfPool*()
+static __TLS int was_called_from_umfPool = 0;
+
+malloc_func_t System_malloc_func;
+calloc_func_t System_calloc_func;
+realloc_func_t System_realloc_func;
+free_func_t System_free_func;
+aligned_alloc_func_t System_aligned_alloc_func;
+
+/******************************************************************************/
+
+void proxy_lib_create_common(void) {
+    umf_os_memory_provider_params_t os_params =
+        umfOsMemoryProviderParamsDefault();
+    enum umf_result_t umf_result;
+
+    umf_result = umfMemoryProviderCreate(umfOsMemoryProviderOps(), &os_params,
+                                         &OS_memory_provider);
+    if (umf_result != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "error: creating OS memory provider failed\n");
+        exit(-1);
+    }
+
+    umf_result = umfPoolCreate(umfJemallocPoolOps(), OS_memory_provider, NULL,
+                               0, &Proxy_pool);
+    if (umf_result != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "error: creating jemalloc pool manager failed\n");
+        exit(-1);
+    }
+}
+
+void proxy_lib_destroy_common(void) {
+    umf_memory_pool_handle_t pool = Proxy_pool;
+    Proxy_pool = NULL;
+    umfPoolDestroy(pool);
+    umfMemoryProviderDestroy(OS_memory_provider);
+}
+
+/******************************************************************************/
+
+static inline void *system_malloc(size_t size) {
+    if (!System_malloc_func) {
```

Comment:
I think there's a race possible here. Shouldn;t proxy_lib_system_alloc_init be called using pthread_init_once?

Reply:
Yes. Fixed.
Done.